fuse_core/src/callback_wrapper.cpp

```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  /**
+   * @brief tell the CallbackGroup how many guard conditions are ready in this waitable
+   */
+  size_t CallbackAdapter::get_number_of_ready_guard_conditions() { return 1;}
+
+  /**
+   * @brief tell the CallbackGroup that this waitable is ready to execute anything
+   */
+  bool CallbackAdapter::is_ready(rcl_wait_set_t * wait_set) {
+    (void) wait_set;
+    return !callback_queue_.empty();
+  }
+
+  /**
+   * @brief add_to_wait_set is called by rclcpp during NodeWaitables::add_waitable() and CallbackGroup::add_waitable()
+    waitable_ptr = std::make_shared<CallbackAdapter>();
+    node->get_node_waitables_interface()->add_waitable(waitable_ptr, (rclcpp::CallbackGroup::SharedPtr) nullptr);
+   */
+  void CallbackAdapter::add_to_wait_set(rcl_wait_set_t * wait_set)
+  {
+    if (RCL_RET_OK != rcl_wait_set_add_guard_condition(wait_set, &gc_, NULL)) {
+      RCLCPP_WARN(rclcpp::get_logger("fuse"), "Could not add callback waitable to wait set.");
+    }
+  }
+
+  /**
+   * @brief check the callback queue and return the next callback to run
+   *
+   */
+  std::shared_ptr< void > CallbackAdapter::take_data(){
+    std::shared_ptr<CallbackWrapperBase> cb_wrapper = nullptr;
+    // fetch the callback ptr and release the lock without spending time in the callback
+    {
+      std::lock_guard<std::mutex> lock(queue_mutex_);
+      if(!callback_queue_.empty()){
+        cb_wrapper = callback_queue_.front();
+        callback_queue_.pop_front();
+      }
+    }
+    return std::static_pointer_cast<void>(cb_wrapper);
+  }
+
+  /**
+   * @brief hook that allows the rclcpp::waitables interface to run the next callback
+   *
+   */
+  void CallbackAdapter::execute(std::shared_ptr<void> & data){
+    if (!data) {
+      throw std::runtime_error("'data' is empty");
+    }
+    std::static_pointer_cast<CallbackWrapperBase>(data)->call();
+  }
+
+  void CallbackAdapter::addCallback(const std::shared_ptr<CallbackWrapperBase> &callback){
+    std::lock_guard<std::mutex> lock(queue_mutex_);
+    callback_queue_.push_back(callback);
+    if (RCL_RET_OK != rcl_trigger_guard_condition(&gc_)) {
+      RCLCPP_WARN(rclcpp::get_logger("fuse"),
+                  "Could not trigger guard condition for callback, pulling callback off the queue.");
+      callback_queue_.pop_back();  // Undo
+    }
+  }
+
+  void CallbackAdapter::addCallback(std::shared_ptr<CallbackWrapperBase> && callback){
+    std::lock_guard<std::mutex> lock(queue_mutex_);
+    callback_queue_.push_back(std::move(callback));
+    if (RCL_RET_OK != rcl_trigger_guard_condition(&gc_)) {
+      RCLCPP_WARN(rclcpp::get_logger("fuse"),
+                  "Could not trigger guard condition for callback, pulling callback off the queue.");
+      callback_queue_.pop_back();  // Undo
+    }
+  }
+
+  void CallbackAdapter::removeAllCallbacks(){
+    std::lock_guard<std::mutex> lock(queue_mutex_);
+    while(!callback_queue_.empty()){
```

Comment:
nit, no need to check if it's emtpy before calling `clear()`, nor loop over clear.

Reply:
https://github.com/locusrobotics/fuse/compare/d3b6c944272008f0ad4f62582c73deb350f3b9e1..3e38f8d4c103da15b673464842e504c4d7a8f799